### ISSUE TITLE
Improve Gas Price UX

### ISSUE THREAD

**Issue by wbobeirne** [MyEtherWallet/MyEtherWallet#571]
The gas price component is one of the more confusing parts of the site:

* It's located in the header, but affects things that happen in the body (Send transaction, interact with contracts.)
* The help text explanation in it, while accurate, is a little verbose. We could probably simplify it, and provide more information via a help link.
* Users cannot see how much ETH / fiat their transaction will cost
* Users cannot set a specific gas price, they're limited by what the slider offers
* Users get no estimate of how long their transaction will take, given the gas they allot
* The default is static and doesn't take the network into account

What I propose we do:

* Create a form component for gas price that is both a slider _and_ a free-form text input
* Place the form component in the send transaction form / contract interact form, remove from header
* Slider should have hard coded defaults, but use [ethgasstation api](https://ethgasstation.info/json/ethgasAPI.json) to come up with better slow / fast / fastest ranges.
* Component should also use the rates reducer for estimating how much the fee will be in fiat

This is currently blocked by `transaction-refactor` (#456) but will probably need some mockups to really get going. @HenryNguyen5 has already shown some progress on calculating eth cost of gas, so he may have ideas for how this should be implemented.

**Comment by tayvano** [comment]
IMO priorities:

1. Integrate with ETH Gas Station to pull recommended gas price

2. Show ETH & USD value of Max TX fee (`gas price * gas limit`)

3. Warn loudly if max tx fee is above X.

The reason it's in the header was to avoid placing it in every component and refactoring the entire thing when we added it. That's literally the only reason so moving it is a good idea.

I'm strongly against a text input for gas price. We have issues as it is. People go to high, they blame us for high fees. If they go to low, they blame us for their TX taking forever. Advanced users can (and do) manually set via inspect element and while not ideal, our userbase is not going to get smarter and we must take steps to protect users from themselves. Moving forward, we are less and less the user and we cannot build a product for ourselves. A text input will undoubtedly lead to people putting 150000 in the gas price field instead of gas limit one, which leads to 1 ETH tx fees, which leads to (essentially) loss. It doesn't help that I hear "increase the gas to 50 GWEI" or "increase the gas to 150000" all the time.

MEW used to set the gas price for the user and doing so was optimal, as long as we updated it with the network changes. The reality is: people don't know what gas is, the difference between gas price and gas limit, and they don't **want** to know these things. Aiming for not even seeing gas price is better than aiming for letting the user have all the options.

**Comment by wbobeirne** [comment]
> Advanced users can (and do) manually set via inspect element

The ability to do this hack is going to go away with the React rewrite. I definitely think if we put in a text field, it should be:
1. Togglable (Default to hidden)
2. Still be contained within our gas send limits, to prevent crazy high transaction costs. Max out at ethgasstation's high number * 1.5, or something to that effect.

**Comment by wbobeirne** [comment]
Here's my first quick crack at what I imagine this would look like. Since we calculate gas limit while online, I think we can rid ourselves of it by default. These don't show help text or validation, but I'm thinking each advanced field would have a ( ? ) button, and we'd have hard coded gas limits that would either error out, or force you to confirm before generating the transaction. The only thing missing in these mockups that I'd like to find out if it's possible to do would be time estimates.

## Default

![send - default](https://user-images.githubusercontent.com/649992/34008685-2f0642bc-e0d4-11e7-94c9-8a550496598a.png)


## Advanced

![send - advanced](https://user-images.githubusercontent.com/649992/34008684-2d62fcac-e0d4-11e7-984f-e7c386e49834.png)


## Offline (Forces advanced due to lack of ethgasstation access)

![send - offline](https://user-images.githubusercontent.com/649992/34008681-2b79c2ae-e0d4-11e7-92bf-83829badde4c.png)

^ Without being able to check ethgasstation, I'd want to make sure we don't have stale hard-coded default ranges.

**Comment by dternyak** [comment]
I think it might even make sense to get rid of the go offline/online terminology and simply show Nonce when advanced. We can "force" advanced when a user doesn't have a connection, and let them toggle nonce visibility when they toggle advanced options.

**Comment by dternyak** [comment]
Overall I like the look a lot 👍

**Comment by dternyak** [comment]
FYI - Integrating with Ethgasstation is out of scope for the MVP - as we will need to having a caching layer to prevent overloading their servers.

**Comment by wbobeirne** [comment]
Sure thing, I'll just stick with the UI for now.

**Comment by tayvano** [comment]
Update from EthGasStation:
https://medium.com/@ethgasstation/gasstation-express-a-simple-gas-price-oracle-for-anyone-running-a-full-ethereum-node-f1bde46260f5

**Comment by wbobeirne** [comment]
To reiterate @dternyak's point

>OK, so if we detect offline (the redux offline boolean), we should
>1) Force advanced mode (Possible add copy to explain this)
>2) Disable the broadcast buton (Change button copy to explain this)

**Comment by dternyak** [comment]
I wanted to get #728 in as part of 0.0.7, but I noticed a few things during a final test that can be improved in a follow-up:

1. When advanced is toggled, show nonce field.
2. Fix prop errors (exception on wallet decrypt).

<img width="702" alt="screen shot 2018-01-07 at 10 46 24 am" src="https://user-images.githubusercontent.com/7861465/34651717-3f7933b6-f399-11e7-80da-10e8bd62ac13.png">

**Comment by dternyak** [comment]
I noticed that the $ amount TX cost isn't shown in the confirmation modal. That would be a good addition IMO.

**Comment by wbobeirne** [comment]
As of #743, the outstanding tasks for this are:

* [ ] Replace Contract Interact call function gas limit with gas slider
    * If you're not sure how to test this, you can try calling a function with bogus data on a Ropsten contract, i.e. ENS Set Owner.
* [ ] Replace Contract Deploy gas limit with gas slider
    * If you're not sure how to test this, you can try deploying a contract on Ropsten, like [your own ERC20 token](https://steemit.com/ethereum/@maxnachamkin/how-to-create-your-own-ethereum-token-in-an-hour-erc20-verified) (You could also use [this module](https://github.com/ConsenSys/Token-Factory))
* [ ] Remove gas slider from header
    * This should only be done after replacing all other gas limit only fields. Except request payment, that stays limit-only.
* [ ] Put gas $ estimate in the confirmation modal

We'll punt on EthGasStation integration for another task.
